pkg/admission/mutator/shoot.go

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2022 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mutator
+
+import (
+	"context"
+	"fmt"
+
+	calicov1alpha1 "github.com/gardener/gardener-extension-networking-calico/pkg/apis/calico/v1alpha1"
+
+	extensionswebhook "github.com/gardener/gardener/extensions/pkg/webhook"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// NewShootMutator returns a new instance of a shoot mutator.
+func NewShootMutator() extensionswebhook.Mutator {
+	return &shoot{}
+}
+
+type shoot struct {
+	decoder runtime.Decoder
+}
+
+// InjectScheme injects the given scheme into the validator.
+func (s *shoot) InjectScheme(scheme *runtime.Scheme) error {
+	s.decoder = serializer.NewCodecFactory(scheme, serializer.EnableStrict).UniversalDecoder()
+	return nil
+}
+
+// Mutate mutates the given shoot object.
+func (s *shoot) Mutate(ctx context.Context, new, old client.Object) error {
+	overlay := &calicov1alpha1.Overlay{Enabled: false}
+
+	shoot, ok := new.(*gardencorev1beta1.Shoot)
+	if !ok {
+		return fmt.Errorf("wrong object type %T", new)
+	}
+
+	networkConfig, err := s.decodeNetworkingConfig(shoot.Spec.Networking.ProviderConfig)
+	if err != nil {
+		return err
+	}
+
+	if old == nil && networkConfig.Overlay == nil {
+		networkConfig.Overlay = overlay
+	}
+
+	if old != nil && networkConfig.Overlay == nil {
+		oldShoot, ok := old.(*gardencorev1beta1.Shoot)
+		if !ok {
+			return fmt.Errorf("wrong object type %T", old)
+		}
+		oldNetworkConfig, err := s.decodeNetworkingConfig(oldShoot.Spec.Networking.ProviderConfig)
```

Comment:
There is a very specific edge case here that blocks shoot deletion. For the current shoots where the networkingConfig is `nil`, you will still try to mutate them but on deletion updates to these fields are not possible. I think you should ignore changes to shoots with termination timestamp